ogame.go

```diff
@@ -2454,6 +2454,272 @@ func (b *OGame) sendFleet(celestialID CelestialID, ships []Quantifiable, speed S
 	return Fleet{}, errors.New("could not find new fleet ID")
 }
 
+// Ensure fleet is sent, returns an error if there's not enough ships
+
+func (b *OGame) ensureFleet(celestialID CelestialID, ships []Quantifiable, speed Speed, where Coordinate,
+	mission MissionID, resources Resources, expeditiontime int) (Fleet, error) {
+
+	// Keep track of start time. We use this value to find a fleet that was created after that time.
+	start := time.Now()
+
+	// Utils function to extract hidden input from a page
+	getHiddenFields := func(pageHTML []byte) map[string]string {
+		doc, _ := goquery.NewDocumentFromReader(bytes.NewReader(pageHTML))
+		fields := make(map[string]string)
+		doc.Find("input[type=hidden]").Each(func(i int, s *goquery.Selection) {
+			name, _ := s.Attr("name")
+			value, _ := s.Attr("value")
+			fields[name] = value
+		})
+		return fields
+	}
+
+	// Page 1 : get to fleet page
+	pageHTML, err := b.getPageContent(url.Values{"page": {"fleet1"}, "cp": {strconv.Itoa(int(celestialID))}})
+	if err != nil {
+		return Fleet{}, err
+	}
+
+	fleet1Doc, _ := goquery.NewDocumentFromReader(bytes.NewReader(pageHTML))
+	fleet1BodyID := fleet1Doc.Find("body").AttrOr("id", "")
+	if fleet1BodyID != "fleet1" {
+		now := time.Now().Unix()
+		b.error(ErrInvalidPlanetID.Error()+", planetID:", celestialID, ", ts: ", now)
+		return Fleet{}, ErrInvalidPlanetID
+	}
+
+	// Ensure we're not trying to attack/spy ourself
+	destinationIsMyOwnPlanet := false
+	myPlanets := ExtractPlanets(pageHTML, b)
+	for _, p := range myPlanets {
+		if p.Coordinate.Equal(where) && p.GetID() == celestialID || (p.Moon != nil && p.Moon.Coordinate.Equal(where) && p.Moon.GetID() == celestialID) {
+			return Fleet{}, errors.New("origin and destination are the same")
+		}
+		if p.Coordinate.Equal(where) || (p.Moon != nil && p.Moon.Coordinate.Equal(where)) {
+			destinationIsMyOwnPlanet = true
+			break
+		}
+	}
+	if destinationIsMyOwnPlanet {
+		switch mission {
+		case Spy:
+			return Fleet{}, errors.New("you cannot spy yourself")
+		case Attack:
+			return Fleet{}, errors.New("you cannot attack yourself")
+		}
+	}
+
+	availableShips := ExtractFleet1Ships(pageHTML)
+
+	enoughShips := true
+	for _, ship := range ships {
+		if ship.Nbr > availableShips.ByID(ship.ID) {
```

Comment:
After I had mentioned this in the Discord chat, I created the following change in my own copy of the source. I noticed it worked fine, but wanted to give it some more time before filing a Pull Request:

```
        for _, ship := range ships {
                if ship.Nbr > 0 && availableShips.ByID(ship.ID) > 0 {
+                       if ship.Nbr > availableShips.ByID(ship.ID) {
+                               return Fleet{}, ErrNotEnoughShips
+                       }
+
                        atLeastOneShipSelected = true
                        break
                }
```

Using this if-statement we also keep the original code working, where I think it checks if a type of ship is available and not only checks the amount of ships. Implementing in into sendFleet saves us from maintaining two separate functions (sendFleet vs ensureFleet). We could also consider a separate parameter (a boolean) to the function to signal whether the sendFleet has to be handled strict or not?

Let met know your thoughts :)

Reply:
> We could also consider a separate parameter (a boolean) to the function to signal whether the sendFleet has to be handled strict or not?

This seems the most efficient way to go about it so we get to keep the functionality of both functions.